### DISCUSSION TITLE
How to create multiple coins?

### DISCUSSION THREAD

**Discussion by 0x-j** [aptos-labs/aptos-developer-discussions#122] [Questions] [answered]
Asked by `ranahardik` on discord `dev-discussion`
---
i want to create multiple coin from one structure  is this possible if it's not so how we can do this ?

like i have one structure for coin once we create coin and use that structure for initialize that so can't use that structure  because thats already on recourse account so what we do in this situation ?

```rust
    struct Bitcoin {}
    
    struct StoreCap has key{
        mint_capability:MintCapability<Bitcoin>,
        burn_capability:BurnCapability<Bitcoin>,
        freeze_capability:FreezeCapability<Bitcoin>,
        // resource_signer_cap:account::SignerCapability
    }

    public entry fun create_coin(owner:&signer){
        let (burn_cap,freeze_cap,mint_cap) = coin::initialize<Bitcoin>(owner,string::utf8(b"Bitcoin"),string::utf8(b"BTC"),8,false);
        move_to(owner,StoreCap{mint_capability:mint_cap, burn_capability:burn_cap,freeze_capability:freeze_cap});
        // coin::register<Bitcoin>(owner);
    }
```
so here when we create token first time it will already move Bitcoin structure in to resource account now if i want to create another coin so we can do this ?

**Comment by 0x-j** [comment]
<del> You are right each account can only hold one instance of a resource of the same type. You need to create a new resource for every coin you create. You can also use [object](https://aptos.dev/move/move-on-aptos/objects/) instead of resource account. Object is a successor to resource account. </del>

**Reply by MD-RanaHardik** [reply]
Do you have any reference that creates coins using this method?

**Comment by 0x-j** [comment]
Calling @gregnazario for help, when i run the unit test in below, it triggers `ECOIN_INFO_ADDRESS_MISMATCH`, description is 
```
Address of account which is used to initialize a coin `CoinType` doesn't match the deployer of module
```
The only way initialize can work is I call it with deployer account and pass deployer signer. I don't understand why we have this restriction, how do we support apps like launchpad where we need one contract as the entry point for people to create coin? 

```rust
module playgrouund_addr::coin_playground {
    use std::string;
    use aptos_framework::account;
    use aptos_framework::coin;

    struct TestCoin has key {}

    struct CoinCap has key {
        mint_capability: coin::MintCapability<TestCoin>,
        burn_capability: coin::BurnCapability<TestCoin>,
        freeze_capability: coin::FreezeCapability<TestCoin>,
    }

    public entry fun create_coin(sender: &signer) {
        let (resource_signer, _signer_cap) = account::create_resource_account(sender, b"token_account");
        let (burn_cap, freeze_cap, mint_cap) = coin::initialize<TestCoin>(
            &resource_signer,
            string::utf8(b"Test"),
            string::utf8(b"symbol"),
            8,
            true
        );

        move_to(
            &resource_signer,
            CoinCap {
                burn_capability: burn_cap,
                freeze_capability: freeze_cap,
                mint_capability: mint_cap
            }
        );
    }

    #[test(sender = @playgrouund_addr, user1 = @0x100, user2 = @0x101)]
    fun test_happy_path(sender: &signer, user1: &signer, user2: &signer) {
        // this will trigger ECOIN_INFO_ADDRESS_MISMATCH in 0x1::coin
        create_coin(sender);
    }
}
```

**Reply by gregnazario** [reply, marked as answer]
Only the publisher of the struct can create a coin with that struct, and they can only do so once. 

To create multiple coins, deploy more structs and use the publisher's account to create the coins

**Reply by gregnazario** [reply]
In test cases you have to use the test coin functions

**Reply by 0x-j** [reply]
> To create multiple coins, deploy more structs and use the publisher's account to create the coins

I don't really understand this part, I have to upgrade the module (so i can add new struct) to create new coin?

Also, is this only the case for coin? I'm looking at this [fungible asset example](https://github.com/aptos-labs/aptos-core/blob/c00b317ff980a0b4d383a78143e9121dc5dd5841/aptos-move/move-examples/fungible_asset/managed_fungible_asset/sources/coin_example.move#L14-L24), it seems like i can move the logic inside initialize to a function called `create_new_fa` and use a counter as seed , then I can keep calling it to create new FAs? This would do what I expected.

**Reply by 0x-j** [reply]
Here's my modified version and it works fine, I can call `create_fa` as many times as i want.

```rust
    fun init_module(sender: &signer) {
        move_to(sender, Counter { count: 0 })
    }
    
    public entry fun create_fa(sender: &signer) acquires Counter {
        let counter = borrow_global_mut<Counter>(@playgrouund_addr);
        let symbol = to_bytes(&string_utils::format2(&b"{}_{}", ASSET_SYMBOL, counter.count));
        let constructor_ref = &object::create_named_object(sender, symbol);
        managed_fungible_asset::initialize(
            constructor_ref,
            0, /* maximum_supply. 0 means no maximum */
            utf8(b"You only live once"), /* name */
            utf8(symbol), /* symbol */
            8, /* decimals */
            utf8(b"http://example.com/favicon.ico"), /* icon */
            utf8(b"http://example.com"), /* project */
            vector[true, true, true], /* mint_ref, transfer_ref, burn_ref */
        );
        counter.count = counter.count + 1;
    }
```

**Reply by MD-RanaHardik** [reply]
did this work as a normal coin? if yes so how we can see this coin in to patra wallet

**Reply by 0x-j** [reply]
[Fungible asset](https://aptos.dev/standards/fungible-asset) is the successor to coin, however DEXes on Aptos haven't fully supported it yet. if you are only playing with the contract you can use FA. Here's the [full code](https://github.com/0xaptosj/playground/blob/main/sources/laucnchpad.move#L35-L50) of my snippet.

**Reply by MD-RanaHardik** [reply]
so for now we have to find other solutions to create the coins

**Reply by Metal-joker** [reply]
lol, i'm also stuck here. Any progress? @0xaptosj @gregnazario Since there's no dex supports Fungible assets so we still need this feature.

**Reply by 0x-j** [reply]
I think short answer is you need to deploy a new contract whenever you want to create a new coin. But let me ask around. 
We have another person asking similar question https://github.com/aptos-labs/aptos-developer-discussions/discussions/146

**Reply by Metal-joker** [reply]
yeah, that's also me, but how aptos gonna support launchpad tho. @0xaptosj

**Reply by 0x-j** [reply]
Ah haha, I found more detail about the approach by pancake swap in the [doc](https://aptos.dev/move/book/generics/#unused-type-parameters). But not sure that approach can work with launchpad. Pancake has this generic Coin
```
struct GenericCoin<phantom T> has key {}
```
You can initialize it by 
```
public entry fun create_coin<T>
coin::initialize<GenericCoin<T>>(...)
```
But the struct type you pass to `create_coin` still needs to exist first, this is fine for pool in AMM DEX because you only create pool with existing coins. But not sure how to do it with launchpad.

**Reply by Metal-joker** [reply]
yes, the first thing come in my mind is how dex create LP tokens too, turns out that they use `LP<X, Y>` and `XY` already exists, I think this can't be used in launchpad. do you think there's any way to create generics?

**Reply by 0x-j** [reply]
I asked internally, unfortunately, there is no other way. I could think of an ugly workaround, that is you upload a new contract with a dummy struct every time someone wants to create a coin. Or you create say 10000 structs beforehand

**Reply by Metal-joker** [reply]
lol, that would be unacceptably ugly, I'll wait till swap supports fungible assest maybe. Is there a hard ddl for coin deprecation or fungible assets support?

**Reply by 0x-j** [reply]
I don't think so, we haven't finalized the migration plan from coin to FA yet 😞, you can look at this [AIP](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-63.md) for detail. Here's the open [PR](https://github.com/aptos-labs/aptos-core/pull/11224).

**Reply by Metal-joker** [reply]
sheesh! this aip seems will take a lot of work, not gonna be done anytime soon I think. or if I started a dex with fully support for fungible assets, will it be backed by Aptos? 😝
